spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/AbstractLocalDeployerSupport.java

```diff
@@ -194,45 +201,61 @@ protected ProcessBuilder buildProcessBuilder(AppDeploymentRequest request, Map<S
 		}
 
 		ProcessBuilder builder = new ProcessBuilder(commands);
+
 		if (!(request.getResource() instanceof DockerResource)) {
-			builder.environment().putAll(appInstanceEnv);
-			builder.environment().putAll(appInstanceEnvToUse);
+			builder.environment().putAll(appPropertiesToUse);
 		}
+
 		retainEnvVars(builder.environment().keySet());
 
 		if (this.containsValidDebugPort(request.getDeploymentProperties(), deploymentId)) {
-			int portToUse = calculateDebugPort(request.getDeploymentProperties(), appInstanceNumber.orElseGet(() -> 0));
-			builder.command().add(1, this.buildRemoteDebugInstruction(
+
+			int portToUse = calculateDebugPort(request.getDeploymentProperties(), appInstanceNumber.orElse(0));
+
+			String debugInstruction = this.buildRemoteDebugInstruction(
 					request.getDeploymentProperties(),
 					deploymentId,
-					appInstanceNumber.orElseGet(() -> 0),
-					portToUse));
+					appInstanceNumber.orElse(0),
+					portToUse);
+
+			if(request.getResource() instanceof DockerResource) {
+				builder.command().add(2, "-e");
+				builder.command().add(3, "JAVA_TOOL_OPTIONS=\""+ debugInstruction + "\"");			}
+			else {
+				builder.command().add(1, debugInstruction);
+			}
 		}
 
+		logger.info(String.format("Command to be executed: %s", String.join(" ", builder.command())));
+
 		return builder;
 	}
 
-	protected void handleAppPropertiesPassing(AppDeploymentRequest request, Map<String, String> appProperties,
-											Map<String, String> appInstanceEnvToUse,
-											Map<String, String> appPropertiesToUse) {
+	protected Map<String, String> formatApplicationProperties(AppDeploymentRequest request,
+											Map<String, String> appInstanceEnvToUse) {
+		Map<String, String> applicationPropertiesToUse =
+				new HashMap<>(appInstanceEnvToUse);
+
 		if (useSpringApplicationJson(request)) {
 			try {
 				//If SPRING_APPLICATION_JSON is found, explode it and merge back into appProperties
-				Map<String, String> localApplicationProperties = new HashMap<>(appProperties);
-				if(localApplicationProperties.containsKey(SPRING_APPLICATION_JSON)){
-					localApplicationProperties.putAll(OBJECT_MAPPER.readValue(appProperties.get(SPRING_APPLICATION_JSON), new TypeReference<HashMap<String,Object>>() {}));
-					localApplicationProperties.remove(SPRING_APPLICATION_JSON);
+				if(applicationPropertiesToUse.containsKey(SPRING_APPLICATION_JSON)){
+					applicationPropertiesToUse.putAll(OBJECT_MAPPER.readValue(applicationPropertiesToUse.get(SPRING_APPLICATION_JSON), new TypeReference<HashMap<String,Object>>() {}));
+					applicationPropertiesToUse.remove(SPRING_APPLICATION_JSON);
 				}
-				appInstanceEnvToUse.putAll(Collections.singletonMap(
-						SPRING_APPLICATION_JSON, OBJECT_MAPPER.writeValueAsString(localApplicationProperties)));
+
+				String saj = OBJECT_MAPPER.writeValueAsString(applicationPropertiesToUse);
+
+				applicationPropertiesToUse = new HashMap<>(1);
+
+				applicationPropertiesToUse.put(SPRING_APPLICATION_JSON, saj);
```

Comment:
Good idea.  However, `Collections.singleMap` returns an unmodifiable `Map`.  The original API modified the `Map` in many places so that wouldn't have worked.  I haven't done a full analysis as to if that still occurs.  I was trying to minimize it as part of this refactor so that API side effects were minimized but I'm not there with 100% confidence.